### PR TITLE
[ISSUE#1881][MAS2.1.2][Keyboard Navigation- Welcome Tab screen] Focus gets trapped on first and last element of the page when user navigates through the page

### PR DESCRIPTION
**Pull request by ceciliaavila**
Solves # 1881

### Description
Fixes how the focus gets trapped between the last and the first element of the pane.
 
### Changes made
We added a new condition to the **_EventHandlers_** class which will manage the focus when the emulator is running in _macOS_. Also, we included two auxiliary methods to find the first and last elements of the active pane and be able to move the focus between them.

Additionally, we added two unit tests to cover the new _keydown_ event.

### Testing
In the following image, you can see how the focus goes from the last element of the pane to the first when navigating with Tab key and vice-versa when using Shift+Tab navigation: 

FocusTrap

### CODE REVIEW

File: packages/app/client/src/utils/eventHandlers.ts [hunk 1]
```diff
@@ -35,13 +35,36 @@ import { Notification, NotificationType, SharedConstants } from '@bfemulator/app
...
+    }
+  }
+
+  public static getLastDecendants(node, list = []) {
+    list = list || [];
+    if (node.children.length > 0) {
+      var child = this.getLastChildWithChildren(node);
+      if (child) {
+        list.push(child);
+        this.getLastDecendants(child, list);
```
Comment (Aliandi) [inline packages/app/client/src/utils/eventHandlers.ts]: Couldn't this all be solved with a map?

Comment (Aliandi) [reply packages/app/client/src/utils/eventHandlers.ts]: `list = node.children.map(function(node) { this.getLastChildwithChildren(node) })`

File: packages/app/client/src/utils/eventHandlers.ts [hunk 2]
```diff
@@ -35,13 +35,36 @@ import { Notification, NotificationType, SharedConstants } from '@bfemulator/app
...
+  public static getLastChildWithChildren(node) {
+    for (let index = node.children.length; index > 0; index--) {
+      if (node.children[index - 1].children.length > 0 && !node.children[index - 1].hidden) {
+        return node.children[index - 1];
+      }
+    }
+  }
+
+  public static getLastDecendants(node, list = []) {
+    list = list || [];
```
Comment (Aliandi) [inline packages/app/client/src/utils/eventHandlers.ts]: Also, why this? list should be either something or the empty list already, this looks redundant

File: packages/app/client/src/utils/eventHandlers.ts (hunk 1 above)
Comment (Aliandi) [reply packages/app/client/src/utils/eventHandlers.ts]: should probably filter after this

File: packages/app/client/src/utils/eventHandlers.ts (hunk 2 above)
Comment (ceciliaavila) [reply packages/app/client/src/utils/eventHandlers.ts]: Feedback applied. Thanks!

File: packages/app/client/src/utils/eventHandlers.ts (hunk 1 above)
Comment (ceciliaavila) [reply packages/app/client/src/utils/eventHandlers.ts]: We tried it but didn't work. Thank you!!

**Comment by Bill7zz** [comment]
I suggest changing `we added a new keydown event for the Tab key` with `we added a new condition to the eventHandlers class which will manage the focus when the emulator is running in Mac OS`

**Comment by ceciliaavila** [comment]
> I suggest changing `we added a new keydown event for the Tab key` with `we added a new condition to the eventHandlers class which will manage the focus when the emulator is running in Mac OS`

Done. Thanks!

File: packages/app/client/src/utils/eventHandlers.spec.ts [hunk 3]
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
     expect(mockRemoteCommandsCalled).toHaveLength(1);
     expect(mockRemoteCommandsCalled[0].commandName).toBe(ToggleDevTools);
   });
+
+  it('should move focus to first element when Tab is pressed', async () => {
+    const event = new KeyboardEvent('keydown', { key: 'Tab' });
+    Object.defineProperty(process, 'platform', { value: 'darwin' });
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.spec.ts]: There'd be a space between variables declarations and the rest of the code.

File: packages/app/client/src/utils/eventHandlers.spec.ts [hunk 4]
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
...
+
+    document.body.innerHTML = mockDOM;
+    var mockFirstElement = document.getElementById('navBtn');
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.spec.ts]: Change both var to const

File: packages/app/client/src/utils/eventHandlers.spec.ts [hunk 5]
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
...
+    var mockLastElement = document.getElementById('btn3');
+    mockLastElement.focus();
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.spec.ts]: add space between line 276 and 277

File: packages/app/client/src/utils/eventHandlers.spec.ts [hunk 6]
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
...
+    await globalHandlers(event);
+
+    expect(document.activeElement.id).toBe(mockFirstElement.id);
+  });
+
+  it('should move focus to last element when Shift+Tab is pressed', async () => {
+    const event = new KeyboardEvent('keydown', { shiftKey: true, key: 'Tab' });
+    Object.defineProperty(process, 'platform', { value: 'darwin' });
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.spec.ts]: Add space here

File: packages/app/client/src/utils/eventHandlers.spec.ts [hunk 7]
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
...
+
+    document.body.innerHTML = mockDOM;
+    var mockFirstElement = document.getElementById('navBtn');
+    var mockLastElement = document.getElementById('btn3');
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.spec.ts]: both var should be const

File: packages/app/client/src/utils/eventHandlers.spec.ts [hunk 8]
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
...
+    mockFirstElement.focus();
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.spec.ts]: add space between line 289 and 290

File: packages/app/client/src/utils/eventHandlers.ts [hunk 9]
```diff
@@ -35,18 +35,41 @@ import { Notification, NotificationType, SharedConstants } from '@bfemulator/app
...
+    for (let index = node.children.length; index > 0; index--) {
+      if (node.children[index - 1].children.length > 0 && !node.children[index - 1].hidden) {
+        return node.children[index - 1];
+      }
+    }
+  }
+
+  public static getLastDecendants(node, list = []) {
+    if (node.children.length > 0) {
+      var child = this.getLastChildWithChildren(node);
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.ts]: change var to const

File: packages/app/client/src/utils/eventHandlers.ts [hunk 10]
```diff
@@ -121,6 +144,27 @@ class EventHandlers {
         } as Notification);
       }
     }
+
+    if (isMac()) {
+      var KEYCODE_TAB = 9;
```
Comment (denscollo) [inline packages/app/client/src/utils/eventHandlers.ts]: All these variables should be const

**Pull request closed without merging**